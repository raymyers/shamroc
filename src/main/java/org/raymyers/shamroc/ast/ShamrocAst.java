package org.raymyers.shamroc.ast;

import java.util.List;
import java.util.Optional;

public interface ShamrocAst {
    record Program(List<Def> defs) {}

    sealed interface Def permits LangDef, SequenceDef {}

    record LangDef(String name, List<LangItem> langItems) implements Def {}
    record SequenceDef(String name) implements Def {}

    sealed interface LangItem permits Record, SumType {}

    record Record(String name, List<RecordField> recordFields) implements LangItem, SumTypeOption {}

    record RecordField(List<NameOrType> nameOrTypes) {}

    record SumType(String name, List<SumTypeOption> sumTypeOptions) implements LangItem {}

    sealed interface SumTypeOption permits Record, NameOrType {}

    record NameOrType(Ident ident1, Optional<Ident> ident2) implements SumTypeOption {}

    record Ident(String string) {}
}
