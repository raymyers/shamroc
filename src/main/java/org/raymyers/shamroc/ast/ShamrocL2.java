package org.raymyers.shamroc.ast;

import java.util.List;
import java.util.Optional;

public interface ShamrocL2 {
    record Program(
            List<LangDef> langDefs,
            List<SequenceDef> sequenceDefs
    ) {}
    record LangDef(String name, List<LangItem> langItems) {}
    record SequenceDef(String name) {}

    sealed interface LangItem permits Record, SumType {}

    record Record(String name, List<RecordField> recordFields) implements LangItem, SumTypeOption {}

    record RecordField(List<TypedIdent> nameOrTypes) {}

    record SumType(String name, List<SumTypeOption> sumTypeOptions) implements LangItem {}

    sealed interface SumTypeOption permits Record, TypedIdent {}

    record TypedIdent(Ident name, Type type) implements SumTypeOption {}

    record Ident(String string) {}

    record Type(String name, List<Type> typeParam) {}
}
