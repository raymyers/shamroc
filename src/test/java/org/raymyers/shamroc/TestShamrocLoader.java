package org.raymyers.shamroc;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.raymyers.shamroc.ast.ShamrocAst;
import org.raymyers.shamroc.ast.ShamrocL2;

import java.io.IOException;
import java.io.InputStream;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class TestShamrocLoader {

    public static final String EXAMPLE_FILE = "Shamroc.shamroc";

    @Test
    void basicParse() throws IOException {
        InputStream stream = this.getClass().getResourceAsStream(EXAMPLE_FILE);
        ShamrocLoader subject = new ShamrocLoader(stream);
//        System.out.println(subject.getSyntaxErrors());
        assertTrue(subject.valid());
        Approvals.verify(subject.getStringTree());
    }

    @Test
    void transformToAst() throws IOException {
        InputStream stream = this.getClass().getResourceAsStream("Shamroc.shamroc");
        ShamrocLoader subject = new ShamrocLoader(stream);
//        System.out.println(subject.getSyntaxErrors());
        assertTrue(subject.valid());
        Approvals.verify(subject.toAst());
    }

    @Test
    void transformToL2() throws IOException {
        InputStream stream = this.getClass().getResourceAsStream("Shamroc.shamroc");
        ShamrocLoader subject = new ShamrocLoader(stream);
//        System.out.println(subject.getSyntaxErrors());
        assertTrue(subject.valid());
        ShamrocAst.Program ast = subject.toAst();
        ShamrocL2.Program l2 = new ShamrocAstToL2().transformProgram(ast);
        Approvals.verify(l2);
    }
}
