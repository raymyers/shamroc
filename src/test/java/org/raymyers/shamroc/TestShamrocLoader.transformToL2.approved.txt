Program[langDefs=[LangDef[name=Shamroc, langItems=[SumType[name=Program, sumTypeOptions=[TypedIdent[name=Ident[string=list], type=Type[name=Def, typeParam=[]]]]], SumType[name=Def, sumTypeOptions=[TypedIdent[name=Ident[string=deflang], type=Type[name=DefLang, typeParam=[]]]]], Record[name=DefLang, recordFields=[RecordField[nameOrTypes=[TypedIdent[name=Ident[string=name], type=Type[name=name, typeParam=[]]], TypedIdent[name=Ident[string=string], type=Type[name=String, typeParam=[]]]]], RecordField[nameOrTypes=[TypedIdent[name=Ident[string=list], type=Type[name=LangItem, typeParam=[]]]]]]], SumType[name=LangItem, sumTypeOptions=[TypedIdent[name=Ident[string=record], type=Type[name=Record, typeParam=[]]], TypedIdent[name=Ident[string=sumtype], type=Type[name=SumType, typeParam=[]]]]], SumType[name=Record, sumTypeOptions=[TypedIdent[name=Ident[string=list], type=Type[name=RecordField, typeParam=[]]]]], Record[name=RecordField, recordFields=[RecordField[nameOrTypes=[TypedIdent[name=Ident[string=nameortype], type=Type[name=NameOrType, typeParam=[]]]]], RecordField[nameOrTypes=[TypedIdent[name=Ident[string=optional], type=Type[name=NameOrType, typeParam=[]]]]]]], Record[name=NameOrType, recordFields=[RecordField[nameOrTypes=[TypedIdent[name=Ident[string=string], type=Type[name=String, typeParam=[]]]]], RecordField[nameOrTypes=[TypedIdent[name=Ident[string=optional], type=Type[name=String, typeParam=[]]]]]]], Record[name=SumType, recordFields=[RecordField[nameOrTypes=[TypedIdent[name=Ident[string=string], type=Type[name=String, typeParam=[]]]]], RecordField[nameOrTypes=[TypedIdent[name=Ident[string=list], type=Type[name=SumTypeOption, typeParam=[]]]]]]], SumType[name=SumTypeOption, sumTypeOptions=[TypedIdent[name=Ident[string=record], type=Type[name=Record, typeParam=[]]], TypedIdent[name=Ident[string=nameortype], type=Type[name=NameOrType, typeParam=[]]]]]]]], sequenceDefs=[]]